[Data][Docs] Convert "Custom datasource" guide to an example (#35751)
ray-project/ray: doc/source/data/examples/custom-datasource.rst

.. _custom_datasources:

================================
Implementing a Custom Datasource
================================

.. note::

  This MongoDatasource guide below is for education only. For production use of MongoDB
  in Ray Data, see :ref:`Creating Dataset from MongoDB <dataset_mongo_db>`.

Ray Data supports multiple ways to :ref:`create a dataset <loading_data>`,
allowing you to easily ingest data of common formats from popular sources. However, if the
datasource you want to read from is not in the built-in list, don't worry, you can implement

> a custom one for your use case. In this guide, we will walk you through how to build

Reviewer: ```suggestion
a custom one for your use case. This guide walks through building
```

> your own custom datasource, using `MongoDB <https://www.mongodb.com/docs/manual/introduction/>`__ as an example.

Reviewer: ```suggestion
a custom datasource, using `MongoDB <https://www.mongodb.com/docs/manual/introduction/>`__ as an example.
```

.. _custom_datasources:

================================
Implementing a Custom Datasource
================================

.. note::

  This MongoDatasource guide below is for education only. For production use of MongoDB
  in Ray Data, see :ref:`Creating Dataset from MongoDB <dataset_mongo_db>`.

Ray Data supports multiple ways to :ref:`create a dataset <loading_data>`,
allowing you to easily ingest data of common formats from popular sources. However, if the

> datasource you want to read from is not in the built-in list, don't worry, you can implement

Reviewer 2: link to the built-in list?

    which expresses document processing in a series of stages (e.g. match documents with a predicate, sort results, and then select a few fields).
    The execution results of the pipelines are used to create dataset.

> A custom datasource is an implementation of :class:`~ray.data.Datasource`. In the

Reviewer: ```suggestion
A custom datasource is an implementation of :class:`~ray.data.Datasource`. In this
```

> example here, let's call it ``MongoDatasource``. At a high level, it will have two

Reviewer: ```suggestion
example, it's called ``MongoDatasource``. At a high level, it has two
```

.. literalinclude:: ./doc_code/custom_datasource.py
    :language: python
    :start-after: __read_single_partition_start__
    :end-before: __read_single_partition_end__

> Once we have this building block, we can just apply it for each of the provided MongoDB

Reviewer: ```suggestion
Once we have this building block, we apply it for each of the provided MongoDB
```

Putting it all together
-----------------------

With ``_MongoDatasourceReader`` and ``_write_multiple_blocks`` above, we are
ready to implement :meth:`create_reader() <ray.data.Datasource.create_reader>`
and :meth:`do_write() <ray.data.Datasource.do_write>`, and put together
a ``MongoDatasource``.

.. literalinclude:: ./doc_code/custom_datasource.py
    :language: python
    :start-after: __mongo_datasource_start__
    :end-before: __mongo_datasource_end__

Now you can create a Dataset from and write back to MongoDB, just like

> any other datasource!

Reviewer: ```suggestion
any other datasource.
```
